test/functional/lib/ipfs-request-gateway-recover.test.js

```diff
@@ -0,0 +1,163 @@
+    beforeEach(function () {
+      state.recoverViaPublicGateway = true
+    })
+    it('should do nothing if broken request is a non-IPFS request', async function () {
+      const request = urlRequestWithStatus('https://ipfs.io', 500)
+      await requestHandler.onCompleted(request)
+      assert.ok(browser.tabs.create.notCalled, 'tabs.create should not be called')
+    })
+    it('should do nothing if broken request is a non-public IPFS request', async function () {
+      const request = urlRequestWithStatus('http://127.0.0.1:8080/ipfs/QmYzZgeWE7r8HXkH8zbb8J9ddHQvp8LTqm6isL791eo14h', 500)
+      await requestHandler.onCompleted(request)
+      assert.ok(browser.tabs.create.notCalled, 'tabs.create should not be called')
+    })
+    it('should do nothing if broken request is to the default public gateway', async function () {
+      const request = urlRequestWithStatus('https://ipfs.io/ipfs/QmYbZgeWE7y8HXkH8zbb8J9ddHQvp8LTqm6isL791eo14h', 500)
+      await requestHandler.onCompleted(request)
+      assert.ok(browser.tabs.create.notCalled, 'tabs.create should not be called')
+    })
+    it('should do nothing if broken request is not a \'main_frame\' request', async function () {
+      const request = urlRequestWithStatus('https://nondefaultipfs.io/ipfs/QmYbZgeWE7y8HXkH8zbb8J9ddHQvp8LTqm6isL791eo14h', 500, 'non_main_frame')
+      await requestHandler.onCompleted(request)
+      assert.ok(browser.tabs.create.notCalled, 'tabs.create should not be called')
+    })
+    it('should redirect broken non-default public gateway IPFS request to public gateway', async function () {
+      const request = urlRequestWithStatus('https://nondefaultipfs.io/ipfs/QmYbZgeWE7y8HXkH8zbb8J9ddHQvp8LTqm6isL791eo14h', 500)
+      await requestHandler.onCompleted(request)
+      assert.ok(browser.tabs.create.withArgs({ url: 'https://ipfs.io/ipfs/QmYbZgeWE7y8HXkH8zbb8J9ddHQvp8LTqm6isL791eo14h', active: true, openerTabId: 20 }).calledOnce, 'tabs.create should be called with IPFS default public gateway URL')
+    })
+  })
+
+  describe('with recoverViaPublicGateway=false', function () {
+    beforeEach(function () {
+      state.recoverViaPublicGateway = false
+    })
+    it('should do nothing on broken non-default public gateway IPFS request', async function () {
+      const request = urlRequestWithStatus('https://nondefaultipfs.io/ipfs/QmYbZgeWE7y8HXkH8zbb8J9ddHQvp8LTqm6isL791eo14h', 500)
+      await requestHandler.onCompleted(request)
+      assert.ok(browser.tabs.create.notCalled, 'tabs.create should not be called')
+    })
+  })
+
+  afterEach(function () {
+    browser.tabs.create.reset()
+  })
+
+  after(function () {
+    delete global.url
+    delete global.browser
+    browser.flush()
+  })
+})
+
+describe('requestHandler.onErrorOccurred:', function () {
+  let state, dnslinkResolver, ipfsPathValidator, requestHandler, runtime
+
+  before(function () {
+    global.URL = URL
+    browser.tabs = { ...browser.tabs, query: sinon.stub().resolves([{ id: 20 }]) }
```

Comment:
Otherwise this call will break this [https://github.com/ipfs-shipyard/ipfs-companion/pull/783/files#diff-79fc41a4f320d6665f9e976b50731ba3R561](call to query).